presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java

```diff
@@ -178,14 +181,41 @@ private Expression unwrapCast(ComparisonExpression expression)
             Type sourceType = typeAnalyzer.getType(session, types, cast.getExpression());
             Type targetType = typeAnalyzer.getType(session, types, expression.getRight());
 
-            if (!hasInjectiveImplicitCoercion(sourceType, targetType)) {
+            ImplicitInjectiveCoercionAbility implicitInjectiveCoercionAbility = getImplicitInjectiveCoercionAbility(sourceType, targetType);
+            boolean boundaryChecksAllowed = implicitInjectiveCoercionAbility == FULL;
```

Comment:
Instead of adding a boolean, I would restructure the code so that there are different methods for different unwrapping routines. It would be easier to follow the flow:
`FULL` --> `out-of-bound unwrapping` --> `in-bound unwrapping` --> give up
`PARTIAL` --> `analyse rightValue` --> either `out-of-bound unwrapping` or `in-bound unwrapping` or NaN or give up